lotus/models/lm.py

```diff
@@ -33,30 +170,72 @@ def format_logprobs_for_cascade(self, logprobs: list) -> tuple[list[list[str]],
-    ) -> list[str] | tuple[list[str], list[dict[str, Any]]]:
-        """Invoke the LLM.
+    # @property
+    # @abstractmethod
+    # def max_ctx_len(self) -> int:
+    #     """The maximum context length of the LLM."""
+    #     pass
 
-        Args:
-            messages_batch (list | list[list]): Either one prompt or a list of prompts in message format.
-            kwargs (dict[str, Any]): Additional keyword arguments. They can be used to specify inference parameters.
+    # @property
+    # @abstractmethod
+    # def max_tokens(self) -> int:
+    #     """The maximum number of tokens that can be generated by the LLM."""
+    #     pass
 
-        Returns:
-            list[str] | tuple[list[str], list[dict[str, Any]]]: A list of outputs for each prompt in the batch. If logprobs is specified in the keyword arguments,
-            then a list of logprobs is also returned.
-        """
-        pass
+    @functools.lru_cache(maxsize=None)
+    def cached_litellm_completion(self,request):
+        return self.litellm_completion(request, cache={"no-cache": False, "no-store": False})
 
-    @property
-    @abstractmethod
-    def max_ctx_len(self) -> int:
-        """The maximum context length of the LLM."""
-        pass
+    def litellm_completion(request, cache={"no-cache": True, "no-store": True}):
+        kwargs = ujson.loads(request)
+        return litellm.completion(cache=cache, **kwargs)
 
-    @property
-    @abstractmethod
-    def max_tokens(self) -> int:
-        """The maximum number of tokens that can be generated by the LLM."""
-        pass
+    @functools.lru_cache(maxsize=None)
+    def cached_litellm_text_completion(self, request):
+        return self.litellm_text_completion(request, cache=None)
+ 
+    def litellm_text_completion(request, cache={"no-cache": True, "no-store": True}):
+        kwargs = ujson.loads(request)
+
+        # Extract the provider and model from the model string.
+        model = kwargs.pop("model").split("/", 1)
+        provider, model = model[0] if len(model) > 1 else "openai", model[-1]
+
+        # Use the API key and base from the kwargs, or from the environment.
+        api_key = kwargs.pop("api_key", None) or os.getenv(f"{provider}_API_KEY")
+        api_base = kwargs.pop("api_base", None) or os.getenv(f"{provider}_API_BASE")
+
+        # Build the prompt from the messages.
+        prompt = '\n\n'.join([x['content'] for x in kwargs.pop("messages")] + ['BEGIN RESPONSE:'])
+
+        return litellm.text_completion(cache=cache, model=f'text-completion-openai/{model}', api_key=api_key,
+                                    api_base=api_base, prompt=prompt, **kwargs)
+
+
+    def _green(text: str, end: str = "\n"):
+        return "\x1b[32m" + str(text).lstrip() + "\x1b[0m" + end
+
+    def _red(text: str, end: str = "\n"):
+        return "\x1b[31m" + str(text) + "\x1b[0m" + end
+
+    def _inspect_history(self,lm, n: int = 1):
```

Comment:
Not a bad idea to credit DSPy here, or at the top of the file, depending on how much we change. You can see how we credit DSPy in `lotus/settings.py` as an example.